### PR TITLE
Editor: Revisions: Diff revisions only once

### PR DESCRIPTION
**Pull request by marekhrabe**
Previously, revisions were diffed on receive time just to get the count of added and removed words, but the exact diff wasn't saved and had to be recomputed again from scratch once the revision was opened. This PR saves the whole diff to state, together with change summary. The counter now also reflects both changes in the content and in the title.

To test:
- open a post with revisions
- open revisions in the sidebar
- make sure you still see "words added, words removed" summaries in the revision list and it takes into count both title and content updates
- switching revisions should have no significant delay, as they are precomputed

### CODE REVIEW

File: client/post-editor/editor-revisions-list/item.jsx [hunk 1]
```diff
@@ -37,30 +37,30 @@ class EditorRevisionsListItem extends PureComponent {
 						} ) }
 				</span>
 				<div className="editor-revisions-list__changes">
-					{ this.props.revision.changes.added > 0 && (
+					{ this.props.revision.summary.added > 0 && (
```
Comment (spen) [inline client/post-editor/editor-revisions-list/item.jsx]: Just a thought here and a little outside the scope of this PR - would it be worth creating variables like so to reduce some noise / improve readibility: 
```js
const hasAddedChanges = this.props.revision.summary.added > 0;
const hasRemovedChanges = this.props.revision.summary.removed > 0;
// ...
{ hasAddedChanges && hasRemovedChanges && ', ' }
```
We could go one step further and use `get()` to as, as it stands, we're accessing chained properties that aren't reflected in the proptypes.

File: client/state/data-layer/wpcom/posts/revisions/index.js [hunk 2]
```diff
@@ -78,9 +94,12 @@ export const receiveSuccess = ( { dispatch }, { siteId, postId }, revisions ) =>
 	const normalizedRevisions = map( revisions, normalizeRevision );
 
 	forEach( normalizedRevisions, ( revision, index ) => {
-		revision.changes = countDiffWords(
-			diffWords( get( normalizedRevisions, [ index + 1, 'content' ], '' ), revision.content )
-		);
+		const previousRevision = get( normalizedRevisions, index + 1, {} );
+		revision.changes = {
```
Comment (spen) [inline client/state/data-layer/wpcom/posts/revisions/index.js]: I like these changes, good work :)

Comment (spen) [reply client/state/data-layer/wpcom/posts/revisions/index.js]: (That's not even a pun)

Comment (spen) [reply client/state/data-layer/wpcom/posts/revisions/index.js]: Also it looks like I'll be able to use `dispatch( withAnalaytics( ... ) )` after this to track the diff generation time :)

File: client/state/data-layer/wpcom/posts/revisions/index.js [hunk 3]
```diff
@@ -4,7 +4,18 @@
...
+	forEach,
+	get,
+	map,
+	mapKeys,
+	mapValues,
+	omit,
+	pick,
+	omitBy,
+	isUndefined,
+} from 'lodash';
```
[the remark this answers is not in the document]
Comment (marekhrabe) [reply client/state/data-layer/wpcom/posts/revisions/index.js]: Are you watching Sorted Food on YouTube or is it a coincidence? 😂

[the remark this answers is not in the document]
Comment (spen) [reply client/state/data-layer/wpcom/posts/revisions/index.js]: coincidence, unfortunately! :P

**Comment by spen** [comment]
Here's some food (unsorted) for thought - how will we eventually handle arbitrary diffing? If we want to diff revision 3 against revision 8, for example, we'll have to do this on the fly.
I wonder if some sort of memoizing would be helpful?

Of course, this is a bridge we can cross when we get to it :)

**Comment by jblz** [comment]
> how will we eventually handle arbitrary diffing?

Not sure re: UI, but we can store it in the state tree under a compound key like `${ fromPostId }_${ toPostId }` & dispatch an action every time a diff is computed.  The list & list item will know what the "base" revision is & select based on that.

**Comment by marekhrabe** [comment]
I've been thinking about it. I would create a new redux store just for changes, next to revisions and store all changes there, comparing any revision to any revision…

Stopping my response right here, as @jblz stole the show with a great explanation :D What I also like about that approach is that changes will be asynchronously computed by design

File: client/state/selectors/test/get-post-revision-changes.js [hunk 4]
```diff
@@ -1,134 +0,0 @@
-/** @format */
```
Comment (jblz) [inline client/state/selectors/test/get-post-revision-changes.js]: @marekhrabe Is there really nothing to test here any longer or did I miss something in the rebase?

**Comment by frontdevde** [comment]
`This branch has conflicts that must be resolved`

**Comment by jblz** [comment]
@frontdevde -- rebased & now with placeholders!

**Pull request merged**